src/modm/platform/clock/samg/clockgen.cpp

```diff
@@ -0,0 +1,159 @@
+/**
+ * Copyright (c) 2021, Jeff McBride
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#include "clockgen.hpp"
+#include <cmath>
+#include <modm/architecture/interface/delay.hpp>
+
+// CMSIS Core compliance
+uint32_t modm_fastdata SystemCoreClock(8000000);
+modm_weak void SystemCoreClockUpdate() { /* Nothing to update */ }
+
+namespace modm::platform
+{
+uint16_t modm_fastdata delay_fcpu_MHz = computeDelayMhz(8'000'000);
+uint16_t modm_fastdata delay_ns_per_loop = computeDelayNsPerLoop(8'000'000);
+
+static void updateSystemClock(uint32_t core_hz) {
+	SystemCoreClock = core_hz;
+	delay_fcpu_MHz = computeDelayMhz(core_hz);
+	delay_ns_per_loop = computeDelayNsPerLoop(core_hz);
```

Comment:
We typically pass the `core_hz` as a template parameter, so these computations happen at compile time (they involve floats).

Reply:
Yeah....to be honest, this occurred to me, and I lazily rationalized away the concern on the basis of, "it's just init code that will run once, I don't care about a few cycles.". But you're right; someone might. I'll template it and the other

Reply:
I wrote a blog post about this, in case you want more info: https://blog.salkinium.com/modm-delay/